### ISSUE TITLE
Scrollbar thumbs aren't selectable when semi-transparent.

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#81015]
When I tried to drag a scrollbar thumb on a desktop app (Linux), grabbing the thumb is pretty hard in the first place, but I was surprised that when I dragged the thumb, it seemed to go the opposite direction from what I expected (the thumb ran away from my mouse pointer), until I realized that the click was instead just going straight through the thumb and the click was being interpreted as a drag of the list contents instead.

Probably when we click on a thumb, it should grab the thumb regardless of whether it is transparent or not, so that it won't be interpreted as a drag of the list contents below it.

https://user-images.githubusercontent.com/8867023/115801150-ea82f080-a390-11eb-8da0-fe7989f2df96.mp4

Sample code from the video (not really specific to this code, however):

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Scroll Bug',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Scroll Bug'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: ListView(
          children: List<Widget>.generate(100, (int index) {
            return Text('Item $index');
          }),
        ),
      ),
    );
  }
}
```

</details>

cc @Piinks

**Comment by pedromassangocode** [comment]
May be a duplicate of https://github.com/flutter/flutter/issues/79235.

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel master, 2.2.0-11.0.pre.245, on macOS 11.2.3 20D91 darwin-x64, locale en-AO)
    • Flutter version 2.2.0-11.0.pre.245 at /Users/pedromassango/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4825c639b6 (15 minutes ago), 2021-04-22 23:26:15 -0700
    • Engine revision 6fa0fb0059
    • Dart version 2.14.0 (build 2.14.0-18.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/pedromassango/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 55.1.5
    • Dart plugin version 211.6693.108

[✓] VS Code (version 1.55.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.3 20D91 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 90.0.4430.85

! Doctor found issues in 1 category.
```

</details>

**Comment by HansMuller** [comment]
CC @Piinks

**Comment by Piinks** [comment]
In the attached video, it looks like the Scrollbar fade out animation has completed by the time the mouse tries to grab it and drag. When it's not visible it's not interactive currently.

https://github.com/flutter/flutter/blob/f55b2665e92aea59645b3b964bc9e2254bff7d1f/packages/flutter/lib/src/widgets/scrollbar.dart#L488

AFAIK users expect when the scrollbar is not visible that it should not interfere with scrolling the scroll view normally, but we could see about making this configurable.

If the scroll view has already scrolled, meaning the scrollbar has metrics it can lay out with, then we could add support for hit testing when it is transparent. I think it would be nice to support it coming back into view no matter what if the mouse hovers over the scrollbar area.
The trouble is that if the scrollbar is not visible, and has not laid out yet, it does not have any information to inform hit testing, or how/where to paint. 

This is really similar to https://github.com/flutter/flutter/issues/80262, which aims to solve the fact that the scrollbar cannot paint without the scroll metrics.

**Comment by gspencergoog** [comment]
If I missed it because it faded out, I feel like it's just REALLY hard to select, then. It very much felt like I grabbed it and then when I moved the mouse, it went in the entirely opposite direction from the way I expected (because I missed it and dragged the background).

It fades out so fast that I can't grab it with the pointer in time, and I feel like people who aren't as aware of the game they have to play to "catch" the thumb will just never get there in time (I'm thinking about my parents trying to use this interface, for instance). It should feel smooth and easy, not like some kind of video game challenge. :-)

Should it maybe not continue to fade out if the mouse is even close to it (like some kind of padding around it)?

**Comment by Piinks** [comment]
I totally get that. Just watching that video looks like a frustrating experience. This is actually the default behavior for scrollbars on chrome/mac, and it can drive me nuts when I try to grab the thumb. 🤣 

The scrollbar was originally padded like you suggested, but we removed it because it caused issues for nearby drag gestures (PR: https://github.com/flutter/flutter/pull/77755 Issue: https://github.com/flutter/flutter/issues/77354 ) I wonder if we can find a happy middle ground to fine tune this, to remove the padding when trying to drag, but add it when evaluating hover.

**Comment by gspencergoog** [comment]
I think it would help a lot to just stop fading out when the pointer is over the scrollbar region, or event in fact to start fading _in_ when it's over the scrollbar. Although I suppose that might be annoying if you were trying to drag the background near the scrollbar. Maybe it only starts fading in again if it's not already transparent?

The fundamental problem, I think, is that these scrollbars were designed for touch screens, and meant to be indicators, not draggable controls. They're too small, they disappear far too easily/quickly, and aren't easily discoverable. How do I really feel? :smile:

**Comment by Piinks** [comment]
> I think it would help a lot to just stop fading out when the pointer is over the scrollbar region, or even in fact to start fading _in_ when it's over the scrollbar. Although I suppose that might be annoying if you were trying to drag the background near the scrollbar. Maybe it only starts fading in again if it's not already transparent? Or maybe it only starts fading in again if you're over the region and a small (100ms?) delay has passed?

I concur! I think this is totally doable. :)

> The fundamental problem, I think, is that these scrollbars were designed for touch screens, and meant to be indicators, not draggable controls. They're too small, they disappear far too easily/quickly, and aren't easily discoverable. 

Oh, this is the new scrollbar designed for desktop by Material Design. 😓

**Comment by xu-baolin** [comment]
Probably the bar should always show on the desk platform by default.

**Comment by Piinks** [comment]
So I've got a fix almost ready for this, it will make the **_hover_** hit testing larger when using a mouse, but not the when actually clicking on it, since that was previously reported and fixed in #77354. It will also make it possible for hover to be detected and the animation to bring the Opacity back up to 1 even when it is transparent - the caveat being that it must have already painted once to know where at least the scrollbar track area is.

#80262 covers the case of when we don't have any data from the scrollable yet. I've taken a couple cracks at it trying to solve both of these at the same time, but have not found a solution for that one yet.

**Comment by gspencergoog** [comment]
That sounds like it will definitely help.
